### ISSUE TITLE
Bad ordering of full alphabetical index o naccented characters

### ISSUE THREAD

**Issue by tfrancart** [NatLibFi/Skosmos#659]
## At which URL did you encounter the problem?

http://erato.sparna.fr/authority/legal-subject-organisation/fr/

## What steps will reproduce the problem?
1. Go to the above URL
2. Scroll down the list : the letter "É" (uppercase E with an accent) is sorted after letters without accent

## What is the expected output? What do you see instead?

Ideally, the expected behavior is that labels beginning with "É" are listed along labels beginning with "E" (in the same list, just as if they were starting with "E").
Or at least the section "É" (accented E) should appear right after the section "E" (E without accent).

I think the whole thing happens in Vocabulary::getAlphabet : [usort($letters, 'strcoll')](https://github.com/NatLibFi/Skosmos/blob/master/model/Vocabulary.php#L465); a Collator should be used here. I am not sure how this function works, and what is the Locale it is using.

My config.inc contains the following LANGUAGE parameter :

```
$LANGUAGES = array(
  'fr' => 'fr_FR.utf8'
);
```

I assume the associated Locale should be used by the sort function ?

**Comment by osma** [comment]
`usort` with `strcoll` as parameter should sort according to the current locale. Do you have the same problem elsewhere? If you disable full alphabetical index, does the index still sort badly?
Skosmos sets the locale to match the UI language, so it should be `fr_FR.UTF8` in your case.

**Comment by tfrancart** [comment]
You are perfectly right.
The order of the full index (not broken down by first letter) is not driven by the sort function but is determined directly from the result of the SPARQL query below. So the problem is not in the sort function in Skosmos but in the ordering in Virtuoso.

I don't think this can be handled at the SPARQL level; the lines from this SPARQL result should probably be reordered using the same function as the one used to generate the alphabet.

```
SELECT DISTINCT ?s ?label ?alabel FROM <http://data.legilux.public.lu/resource/authority/legal-subject-organisation/graph>
WHERE {
  {
    ?s skos:prefLabel ?label .
    FILTER (
      regex(str(?label), '^.*$', 'i') && langMatches(lang(?label), 'fr')
    )
  }
  UNION
  {
    {
      ?s skos:altLabel ?alabel .
      FILTER (
        regex(str(?alabel), '^.*$', 'i') && langMatches(lang(?alabel), 'fr')
      )
    }
    {
      ?s skos:prefLabel ?label .
      FILTER (langMatches(lang(?label), 'fr'))
    }
  }
  ?s a ?type .
  FILTER NOT EXISTS { ?s owl:deprecated true }
  VALUES (?type) { (<http://www.w3.org/2004/02/skos/core#Concept>) }
}
ORDER BY STR(LCASE(COALESCE(?alabel, ?label)))
```

**Comment by osma** [comment]
We generally avoid reordering SPARQL result sets in application code. That quickly leads to performance issues, in particular it makes paging of results impossible. For example, the alphabetical index is normally loaded 250 entries at a time IIRC.

The problem is that SPARQL endpoints are not very good at collating results in a sensible order. We have issue #559 about this particular problem with the alphabetical index. @kinow has implemented locale-specific collation in Jena (see [JENA-1313](https://issues.apache.org/jira/browse/JENA-1313)), which is included in recent Jena and Fuseki releases (3.4.0 and 3.5.0), and proposed a PR #618 with the changes required on the Skosmos side, though unfortunately the work on merging that PR has stalled.

Do you know if Virtuoso can do language/locale specific collation of results? If it does, then a fix like PR #618 should be possible for Virtuoso as well. Locale-specific collation is unfortunately not standardized in SPARQL, so each triple store requires custom support to enable it.

**Comment by tfrancart** [comment]
Closing this issue since this is Virtuoso-related.
